### ISSUE TITLE
Gradient mask unexpected behavior

### ISSUE THREAD

**Issue by assaft** [aferrero2707/PhotoFlow#138]
Continuous build: photoflow-20170222

I'm trying to set the gradient line in order to mask the upper-left part of the image, image, but the mask I'm getting has some weird fluctuations. In addition, any touch to the point on the right side change it in an unexpected way.

![mask](https://cloud.githubusercontent.com/assets/4903680/23332472/b0110ffa-fb82-11e6-9f49-079311fb7fcf.png)

[Original file](https://dl.dropboxusercontent.com/u/21642925/PhotoFlow/sources/20170224_094753_HDR.jpg)

[PFI file](https://dl.dropboxusercontent.com/u/21642925/PhotoFlow/reports/2017_02_25-1/20170224_094753_HDR.pfi)

**Comment by aferrero2707** [comment]
Could you send me a .pfi file with this kind of weird behaviour? I cannot reproduce this here...

**Comment by assaft** [comment]
The PFI is attached below the picture, isn't it?

**Comment by aferrero2707** [comment]
Sorry, I must be blind!!!

**Comment by assaft** [comment]
Did you manage to reproduce the problem?

**Comment by aferrero2707** [comment]
Yes. The problem is due to the third and fourth points from the right being exactly on the same vertical line.

**Comment by assaft** [comment]
But look at this, here they are clearly not on the same vertical line:
![output](https://cloud.githubusercontent.com/assets/4903680/23340418/97f2440e-fc3e-11e6-980f-fcc7c8444399.png)

[PFI file](https://dl.dropboxusercontent.com/u/21642925/PhotoFlow/reports/2017_02_25-1/20170224_094753_HDR_2.pfi)

**Comment by aferrero2707** [comment]
Well, in fact the controls points of the gradient tool, at least in their current implementation, are ok for introducing some small modulation of the gradient itself, but not for "extreme" changes that locally turn an horizontal gradient into a vertical one... I might try to improve this in the fugure, but at the moment this is not possible.

However, in cases like the one you are showing, one can get very good results by combining a gradient and a patch mask:

<img width="1280" alt="screen shot 2017-02-26 at 16 48 05" src="https://cloud.githubusercontent.com/assets/6229237/23341814/1ffe7330-fc4f-11e6-842f-1095292fc8bf.png">

<img width="1280" alt="screen shot 2017-02-26 at 16 48 23" src="https://cloud.githubusercontent.com/assets/6229237/23341820/2ab1ca5c-fc4f-11e6-82cc-699df4bea61b.png">

[PFI file](https://filebin.net/h3bj4cejl4at1bsj)

**Comment by assaft** [comment]
I have some better understanding of the limitation now (it seems like PhF relies on the slope between points on the gradient line and therefore vertical or close-to-vertical points create problems) but I think that this behavior will be puzzling for new users, as it was for me. 
From the way the gradient tool is presented in the UI - i.e. with the flexibility of moving the points on the gradient line - the user would expect to be able to raise and lower some points to different extents without anticipating the weird fluctuations. 
I find it not intuitive also because it's unclear why the slope matters at all. Given that the user sets the type of the gradient to be Vertical, and the direction is set (i.e. from top to bottom), and the line is explicitly set by the points, and the transition smoothness is set (by the curve on the settings panel) then the gradient mask can be defined unambiguously. I guess PhF is trying to do some smart things behind the scenes but I'm not sure it's needed here.
